src/mongocxx/test/collection.cpp

```diff
@@ -2808,4 +2812,145 @@ TEST_CASE("Ensure that the WriteConcernError 'errInfo' object is propagated", "[
     REQUIRE(contains_err_info);
 }
 
+TEST_CASE("expose writeErrors[].errInfo", "[collection]") {
+    // A helper for checking that an error document is well-formed according to our requirements:
+    auto writeErrors_well_formed = [](const bsoncxx::document::view& reply_view) -> bool {
+        if (!reply_view["writeErrors"]) {
+            return false;
+        }
+
+        const auto& errdoc = reply_view["writeErrors"][0];
+
+        auto error_code = errdoc["code"].get_int32();
+
+        // The code should always be 121 (DocumentValidationFailure):
+        if (121 != error_code) {
+            std::ostringstream os;
+            os << "writeErrors expected to have code 121, but had " << error_code << " instead";
+            throw std::runtime_error(os.str());
+        }
+
+        // We require the "details" field be present:
+        if (!reply_view["writeErrors"][0]["errInfo"]["details"]) {
+            throw std::runtime_error("no \"details\" field in \"writeErrors\"");
+        }
+
+        return true;
+    };
+
+    // Set up our test environment:
+    instance::current();
+
+    auto client_opts = mongocxx::options::client();
+
+    // Listen to the insertion-failed event: we want to get a copy of the server's
+    // response so that we can compare it to the thrown exception later:
+    mongocxx::options::apm apm_opts;
+    apm_opts.on_command_failed([&writeErrors_well_formed](const events::command_failed_event& ev) {
+        writeErrors_well_formed(ev.failure());
```

Comment:
...although I actually think we should never see this? That is, in this particular set of tests, getting a command_failed would be an test failure?

Reply:
Write errors are not considered a command failure. A server reply with `ok: 0` is considered a command failure. A reply with writeErrors still has `ok: 1`:

```
{
  n: 0,
  electionId: ObjectId("7fffffff0000000000000041"),
  opTime: { ts: Timestamp({ t: 1645749523, i: 1 }), t: Long("65") },
  writeErrors: [
    {
      index: 0,
      code: 121,
      errInfo: {
        failingDocumentId: ObjectId("621825131c6f87414554373c"),
        details: {
          operatorName: '$type',
          specifiedAs: { x: { '$type': 'string' } },
          reason: 'type did not match',
          consideredValue: 1,
          consideredType: 'int'
        }
      },
      errmsg: 'Document failed validation'
    }
  ],
  ok: 1,
  '$clusterTime': {
    clusterTime: Timestamp({ t: 1645749523, i: 2 }),
    signature: {
      hash: Binary(Buffer.from("0000000000000000000000000000000000000000", "hex"), 0),
      keyId: Long("0")
    }
  },
  operationTime: Timestamp({ t: 1645749523, i: 1 })
}
```

Listening for a `CommandFailed` event is not a requirement of the test. Only `CommandSucceeded`:

> Additionally, assert that a CommandSucceededEvent was observed and
that the ``writeErrors[0].errInfo`` field in the response document matches the
WriteError's ``details`` property.

Reply:
+1 agreed!